### ISSUE TITLE
sisl.nodes and automatic_recalc

### ISSUE THREAD

**Issue by zerothi** [zerothi/sisl#538]
**Describe the feature**

Instead of passing the automatic_recalc as a function argument, wouldn't it be more explicit to use a separate method for the class?

I.e. a user might find it weird that an _extra_ argument to the function call will change the behaviour.
Since the `update_inputs` is already documented, it shouldn't be too much of a hazzle to make a `Node.set_autocalc(True|False)` or similar? Or perhaps, since the documentation emphasises `lazy` behaviour, we might name it `Node.set_lazy(True|False)`?

Is there a special point on adding extra arguments to the function?

@pfebrer

**Comment by pfebrer** [comment]
I agree the current approach doesn't make much sense.

Let's call this extra arguments that define the behavior of the node "meta-arguments".

In the tutorial I show how one can overwrite a meta-argument default in the function signature:

```python
@Node.from_func
def myalert(val=val, automatic_recalc=True):
    ...
```

The function itself doesn't use `automatic_recalc`, so indeed it is very strange that it has this extra argument. It makes even less sense if the function is defined separately, and you create a node like:

```python
def myalert(val=val, automatic_recalc=True):
    ...

alert_node = Node.from_func(myalert)
```

So for this case, I would propose:
```python
@Node.from_func(automatic_recalc=True)
def myalert(val=val):
    ...

#or

def myalert(val=val):
    ...

Node.from_func(myalert, automatic_recalc=True)
```

To overwrite meta-argument defaults.

Then `lazy` could also be a meta-argument.

I also agree 100% that the `update_inputs` should be left only for function arguments as well. I like your idea of `set_*` methods, but perhaps to make it more general for any number of meta-arguments there could be a special method to update them (`update_metainputs` or whatever, I don't know what the right name would be to make it understandable).

So, in summary, you made me realise that we should have this distinction between `arguments` and `meta-arguments` and treat them separately, do you agree?

**Comment by pfebrer** [comment]
Another meta-argument that I already thought we should introduce is `cached=True|False`. Because you might want a lazy node but still not cache its result.

**Comment by pfebrer** [comment]
I forgot you also made the point that calling the node with this extra arguments is strange. I also agree. How about:

```python
with node_arguments(lazy=True, automatic_recalc=True):
     node = node_cls(argument1="something")

#or

node_cls.with(lazy=True, automatic_recalc=True)(argument1="something")
```

**Comment by zerothi** [comment]
Yeah, the `node_cls.with` seems like a good idea... Then everything can be packed into that one.

```python
def = sisl.node.context(**default args)
def2 = sisl.node.context(**default args2)
node.with(context=def)(...)
node(...)
```
or something similar?

**Comment by pfebrer** [comment]
Do you think the `with` needs to accept everything packed into the `context` argument? Why not pass the arguments directly? Do you expect `with` to have other arguments?

The way you did it is interesting though. Do you expect these settings to be put in a context and then nodes have to check the context that they are attached to before actually doing things? As opposed to the settings being part of each node individually. If you did that you could share the context between nodes, which could be nice to switch on and off some features for multiple nodes at the same time. And by default nodes would point to the main general context.

The only problem that I see is that perhaps I want some extra context setting that only applies to a certain class of Node. But then these contexts could be made of partial contexts that are chained in something like a chainmap:

```python
# All the settings of the main sisl context
sisl_main_context = sisl.nodes.context( ... )

# Special context that I define for event listeners
listen_context = sisl.nodes.context(automatic_recalc=True}

# Use the sisl main context except for the settings that are explicitly overwritten by listen_context 
context = Chainmap(sisl_main_context, listen_context)

node.with(context)()
```

Then any update to the main context of `sisl` would still be applied to the node.

I really like the idea that with one switch in this case for example you could turn off all event listeners. And since we are trying to be functional it seems like having contexts that are consumed by nodes is better than having a node state.

Following the same mindset, input management could also be something detached from nodes themselves, but that is for another discussion :sweat_smile:

**Comment by zerothi** [comment]
My main motivation for having a context argument is that context managers would have problems if one of the nodes in the context do not want to use the context. That would mean an irregular workflow...

That is why the argument based contexts seemed a bit more natural to me. But you could do both, a global *state context*, and then local contexts which are chain-mapped internally for their use. A context should then also be able to revert to default arguments.
